api/tests/unit/application/certification-courses/index_test.js

```diff
@@ -113,6 +112,30 @@ describe('Unit | Application | Certifications Course | Route', function () {
       // then
       expect(response.statusCode).to.equal(200);
     });
+
+    it('return forbidden access if user has METIER role', async function () {
+      // given
+      sinon.stub(securityPreHandlers, 'checkUserHasRoleMetier').callsFake((request, h) => h.response(true));
+      sinon
+        .stub(securityPreHandlers, 'checkUserHasRoleSuperAdmin')
+        .callsFake((request, h) => h.response({ errors: new Error('forbidden') }).code(403));
+      sinon
+        .stub(securityPreHandlers, 'checkUserHasRoleSupport')
+        .callsFake((request, h) => h.response({ errors: new Error('forbidden') }).code(403));
+      sinon
+        .stub(securityPreHandlers, 'checkUserHasRoleCertif')
+        .callsFake((request, h) => h.response({ errors: new Error('forbidden') }).code(403));
+
+      const httpTestServer = new HttpTestServer();
+      await httpTestServer.register(moduleUnderTest);
+
+      // when
+      const response = await httpTestServer.request('PATCH', '/api/certification-courses/1234');
+
+      // then
+      expect(response.statusCode).to.equal(403);
+      sinon.assert.notCalled(securityPreHandlers.checkUserHasRoleMetier);
+    });
```

Comment:
**question**
:thinking: Mais qu'est-ce qu'on teste ici ? 
Ca ressemble plus à un test d'intégration de tous les composants qu'à un test unitaire.

Je m'explique: l'implémentation est un appel à la fonction `userHasAtLeastOneAccessOf` avec une liste de paramètres
Cette fonction est testée en unitaire dans `api/tests/unit/application/security-pre-handlers_test.js`

On veut tester qu'on appelle `userHasAtLeastOneAccessOf` avec les bon paramètres (`checkUserOwnsCertificationCourse`,..)

Je propose un truc du genre (j'ai une 500 même avec l'implém de `_replyForbiddenError` , mais ça doit pouvoir se régler). Comme ça, peu importe que la fonction `userHasAtLeastOneAccessOf` soit refactorée, les tests protègent toujours ?

``` js
    it('return forbidden access if user does not have any of allowed roles', async function () {
      // given
      sinon.stub(securityPreHandlers, 'userHasAtLeastOneAccessOf').withArgs([
                securityPreHandlers.checkUserOwnsCertificationCourse,
                securityPreHandlers.checkUserHasRoleSuperAdmin,
                securityPreHandlers.checkUserHasRoleCertif,
                securityPreHandlers.checkUserHasRoleSupport,
	].callsFake((request, h) => _replyForbiddenError(h));
      const httpTestServer = new HttpTestServer();
      await httpTestServer.register(moduleUnderTest);

      // when
      const response = await httpTestServer.request('PATCH', '/api/certification-courses/1234');

      // then
      expect(response.statusCode).to.equal(403);
    });
```

Reply:
Hé bien c'est initialement ce qu'on avait essayé sans succès 😅  ... si tu arrives à corriger la 500 on est preneur !

Reply:
Au lieu de `.callsFake((request, h) => _replyForbiddenError(h))`, il faudrait ceci `.callsFake(() => (request, h) => _replyForbiddenError(h))` et ça devrait fonctionner.

Reply:
L'erreur 500 contient le message d'erreur qui aiguille sur la piste (response is undefined)
Ca donne ça au final

```js
    it('return forbidden access if user does not have any of allowed roles', async function () {
      // given
      sinon
        .stub(securityPreHandlers, 'userHasAtLeastOneAccessOf')
        .withArgs([
          securityPreHandlers.checkUserOwnsCertificationCourse,
          securityPreHandlers.checkUserHasRoleSuperAdmin,
          securityPreHandlers.checkUserHasRoleCertif,
          securityPreHandlers.checkUserHasRoleSupport,
        ])
        .callsFake(() => {
          return async (request, h) => {
            return _replyForbiddenError(h);
          };
        });
      const httpTestServer = new HttpTestServer();
      await httpTestServer.register(moduleUnderTest);

      // when
      const response = await httpTestServer.request('PATCH', '/api/certification-courses/1234');

      // then
      expect(response.statusCode).to.equal(403);
    });
```

Reply:
`return forbidden access if user has role METIER`